presto-spi/src/main/java/com/facebook/presto/spi/function/LambdaDescriptor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import com.facebook.presto.common.Subfield;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
+public class LambdaDescriptor
+{
+    /**
+     * Index of the argument in the Call expression of the lambda function that this LambdaDescriptor represents
+     * Value -1 indicates that the function does not accept the functional argument.
+     */
+    final int callArgumentIndex;
+    /**
+     * The mapping of the index of lambda argument to the index of the argument of the CallExpression that this argument relates to.
+     * <p>
+     * Example: ZIP_WITH(array_column1, array_column2, (x,y) -> x.quantity * y.price).
+     * The purpose of this mapping is to correctly attribute that subfield `quantity` relates to the `array_column1`, while subfield `price` relates to the `array_column2`.
+     * Here, the lambda argument 0 with the name x relates to the argument 0 in CallExpression that represent ZIP_WITH function, while the lambda argument 1 corresponds
+     * to the argument 1 of the CallExpression. Therefore, the mapping for this example must be as follows:
+     * { 0 => 0, 1 => 1 }.
+     */
+    final Map<Integer, Integer> lambdaArgumentToCallArgumentIndexMap;
+
+    /**
+     * Contains the transformation function that needs to be applied to the lambda subfields of this function.
+     * This parameter must be non-null if function signature accepts the lambda parameter and must be null. Otherwise, if function signature does not accept the lambda
+     * parameter. If transformation succeeded, then the returned value contains the transformed set of lambda subfields. Otherwise, the function must return
+     * <code>Optional.empty()</code> values.
+     */
+    private final Function<Set<Subfield>, Optional<Set<Subfield>>> lambdaSubfieldsTransform;
```

Comment:
`outputToInputTransformationFunction` in `ComplexTypeFunctionDescriptor` and `lambdaSubfieldsTransform` in  `LambdaDescriptor`play different roles.  
In `ComplexTypeFunctionDescriptor`, `outputToInputTransformationFunction` is the inverse that of the  transformation that function does over input. 
In `LambdaDescriptor`, `lambdaSubfieldsTransform` is a transformation between the lambda argument (e.g. the  array element) and the input (e.g. the array).

Reply:
how about this name `lambdaArgumentToInputTransformationFunction`?

Reply:
Renamed `lambdaSubfieldsTransform ` to `lambdaArgumentToInputTransformationFunction `

Reply:
i see won't this always be the same then for every function? it operates on the elements of the thing it's iterating over?

Reply:
I believe that I introduced it because of the `maps` where we have `key` and `value` that would require different transformation functions. Currently, we do not support map keys. However, I am thinking to add it in order to support `map_agg` function.